File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        
      int cnt=0;  
        
      const int n=s.size();
        
        for(int i=0;i<n;i++){
            if(s[i]=='(') cnt++;
             else if(cnt > 0 && s[i]==')') cnt--;
             else if(cnt==0 && s[i]==')') s[i]='*'; 
        }
        cnt=0;
         for(int i=n-1;i>=0;i--){
            if(s[i]==')') cnt++;
             else if(cnt > 0 && s[i]=='(') cnt--;
             else if(cnt==0 && s[i]=='(') s[i]='*'; 
        }
        
        string res="";
        for(int i=0;i<n;i++){
            if(s[i]!='*') res.push_back(s[i]);
            
        }
        
        return res;
    }
};

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
      
        StringBuilder ss = new StringBuilder(s);
        Stack<Integer> st = new Stack<>();
        
        
        for(int i=0;i<ss.length();i++){
            
              if(ss.charAt(i)=='(')
                  st.push(i);
              else if(ss.charAt(i)==')')
              {
                  if(st.isEmpty())
                      ss.setCharAt(i,'*');
                  else
                      st.pop();
              }
        }
        
        while(!st.isEmpty()){
            ss.setCharAt( st.pop() ,'*');
        }
        
        StringBuilder ans = new StringBuilder();
        
        for(int i=0;i<ss.length();i++)
        {
            if(ss.charAt(i)=='*') continue;
            
            ans.append(ss.charAt(i));
        }
        
        
        return ans.toString();
        
    }
}